### ISSUE TITLE
[feature] Implement a generic physics-driven loss class to enforce the governing PDEs

### ISSUE THREAD

**Issue by abdulrahman-1212** [EvertBunschoten/MLPCpp#21]
## Description
This issue proposes the addition of a new, generic `CPhysicsLoss` class to the framework. The goal is to provide a unified, lambda-based interface for users to enforce arbitrary governing Partial Differential Equations (PDEs) during Physics-Informed Neural Network (PINN) training, without needing to write custom C++ loss classes for every new physical system.

## Motivation
Currently, incorporating a new set of governing PDEs into the training loop requires users to write a bespoke C++ class inheriting from `CBaseLoss`. This involves significant boilerplate code for managing variable lookups, iterating over collocation points, and computing the mean squared error of the residuals. 

A generic, mathematically expressive physics loss class will drastically lower the barrier to entry for defining new PDE systems, reduce boilerplate, and centralize the residual computation logic.

## Proposed Design

The new feature consists of three main components:

### 1. `PhysicsState` (Semantic Accessor)
A lightweight wrapper around the network's `PredictionResult`. 
* **Performance:** It resolves string variable names to array indices *once* during the `CPhysicsLoss` construction phase. This ensures zero string-lookup overhead inside the hot evaluation loop.
* **API:** Provides clean, semantic accessors: `In(idx)`, `Out(idx)`, `Jac(out_idx, in_idx)`, and `Hess(out_idx, in1_idx, in2_idx)`.

### 2. `ResidualFunction` (User-Defined PDE Logic)
A `std::function` signature that takes a `PhysicsState` and returns a `std::vector<mlpdouble>`. 
* Users write a lambda that defines the raw (unsquared) residuals for their governing equations. 
* The framework automatically handles squaring and summing these residuals to compute the final Mean Squared Error (MSE) loss.

### 3. `CPhysicsLoss` (The Core Class)
Inherits from `CBaseLoss` and orchestrates the evaluation. 
* **Generality:** Supports single- or multi-equation systems (coupled or uncoupled).
* **Robustness:** Validates at runtime that the user's lambda returns exactly the expected number of residuals (`n_equations`), throwing a clear error if there is a mismatch.
* **Autodiff Sparsity Flags:** Introduces `needs_jacobian` (default `true`) and `needs_hessian` (default `false`) constructor flags. This allows the `CMLPTrainer` to skip expensive reverse-mode autodiff sweeps for losses that only require algebraic constraints (e.g., $p \ge 0$) or only first-order derivatives.

## Usage Example

Here is how a user would define the 2D incompressible Navier-Stokes equations (Continuity + X-Momentum) using the new API:

```cpp
// 1. Define the residual lambda
ResidualFunction navier_stokes_fn = [](const PhysicsState& s) -> std::vector<mlpdouble> {
    // Read inputs
    mlpdouble u = s.Out(0);
    mlpdouble v = s.Out(1);
    mlpdouble p = s.Out(2);

    // Read first derivatives (Jacobian)
    mlpdouble du_dx = s.Jac(0, 0);
    mlpdouble dv_dy = s.Jac(1, 1);
    mlpdouble du_dt = s.Jac(0, 2); // assuming 't' is input index 2
    mlpdouble dp_dx = s.Jac(2, 0);

    // Read second derivatives (Hessian) for viscous term
    mlpdouble d2u_dx2 = s.Hess(0, 0, 0);
    mlpdouble d2u_dy2 = s.Hess(0, 1, 1);

    // Compute raw residuals
    mlpdouble r_continuity = du_dx + dv_dy;
    mlpdouble r_momentum_x = du_dt + u * du_dx + v * du_dy + dp_dx - (d2u_dx2 + d2u_dy2);

    // Return raw residuals (DO NOT pre-square them)
    return { r_continuity, r_momentum_x };
};

// 2. Register the loss with the trainer
auto physics_loss = std::make_shared<CPhysicsLoss>(
    "NavierStokes2D",                   // Loss name
    navier_stokes_fn,                   // Residual lambda
    2,                                  // Number of equations (n_equations)
    {"x", "y", "t"},                    // Inputs used by the equations
    {"u", "v", "p"},                    // Outputs used by the equations
    net.GetInputVars(),                 // Full network input names
    net.GetOutputVars(),                // Full network output names
    true,                               // needs_jacobian
    true                                // needs_hessian (required for viscous terms)
);

trainer.AddLoss(physics_loss);
